### PR TITLE
ROC-5168 Updating case number on PayHub

### PR DESCRIPTION
**Pull request by karthikkannankainos**
### JIRA link
https://tools.hmcts.net/jira/browse/ROC-5168

### Change description
Updating case number on PayHub

### CODE REVIEW

File: src/main/features/claim/routes/pay.ts [hunk 1]
```diff
@@ -94,12 +97,15 @@ async function successHandler (res, next) {
...
+      savedClaim = await claimStoreClient.saveClaim(draft, user, 'admissions')
     } else {
-      await claimStoreClient.saveClaim(draft, user)
+      savedClaim = await claimStoreClient.saveClaim(draft, user)
     }
   }
+  const payClient: PayClient = await getPayClient(req)
+  const paymentReference = draft.document.claimant.payment.reference
 
+  payClient.update(user, paymentReference, savedClaim.externalId, savedClaim.claimNumber)
```
Comment (KiranVarma02) [inline src/main/features/claim/routes/pay.ts]: i think you should have await start if this line...intellij should complain the way it is now

File: src/main/app/pay/payClient.ts [hunk 2]
```diff
@@ -104,23 +116,29 @@ export class MockPayClient implements PayClient {
       })
   }
 
+  async update (user: User, paymentReference: string, caseReference: string, caseNumber: string): Promise<void> {
+    const payUpdateDelayInMs = 681
+    await delay(payUpdateDelayInMs)
+
```
Comment (KiranVarma02) [inline src/main/app/pay/payClient.ts]: i am not sure if above 2 lines are necessary..

Comment (robertparkinson) [reply src/main/app/pay/payClient.ts]: Why the delay?

File: src/main/app/pay/payClient.ts [hunk 3]
```diff
@@ -151,10 +169,41 @@ export class GovPayClient implements PayClient {
...
   }
 
-  private preparePaymentRequest (caseReference: string, externalId: string, fees: Fee[]): object {
+  async update (user: User, paymentReference: string, caseReference: string, caseNumber: string): Promise<void> {
+    checkDefined(user, 'User is required')
+    checkNotEmpty(paymentReference, 'Payment reference is required')
+    checkNotEmpty(caseReference, 'Case Reference is required')
+    checkNotEmpty(caseNumber, 'Case Number is required')
+    try {
+      await request.patch({
```
Comment (robertparkinson) [inline src/main/app/pay/payClient.ts]: Why patch? Just curious if there is a specific need or preference? I don't mind as long as it works :)

File: src/main/app/pay/payClient.ts (hunk 2 above)
Comment (karthikkannankainos) [reply src/main/app/pay/payClient.ts]: I have seen in other mock implementation too, they mimic certain milli sec of delay to mock delay the delay in response. I did the same.

File: src/main/app/pay/payClient.ts (hunk 3 above)
Comment (karthikkannankainos) [reply src/main/app/pay/payClient.ts]: @robertparkinson PAY supports update of existing resource thru PATCH.

Comment (KiranVarma02) [reply src/main/app/pay/payClient.ts]: I am fine if it works too but we used request.put in other cases in our project. We haven't used patch anywhere. Let me know what you think!

File: src/main/app/mocks/mockPayClient.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  async update (user: User, paymentReference: string, caseReference: string, caseNumber: string): Promise<void> {
+    const payUpdateDelayInMs = 681
+    await delay(payUpdateDelayInMs)
+
+    return Promise.resolve()
+  }
+
+}
```
Comment (KiranVarma02) [inline src/main/app/mocks/mockPayClient.ts]: Why is this file in main folder? i think it should be moved to test..

Comment (karthikkannankainos) [reply src/main/app/mocks/mockPayClient.ts]: @KiranVarma02 this is required as this mock pay is used by Performance test environment, so it cant be in test folder, it has to be in a folder where we can deploy the code. Have already discussed with @robertparkinson and explained.

File: src/main/app/pay/payClient.ts [hunk 5]
```diff
@@ -151,10 +95,41 @@ export class GovPayClient implements PayClient {
...
+        }
+      })
+      return
+    } catch (err) {
+      if (err.statusCode === HttpStatus.NOT_FOUND) {
+        return undefined
+      }
+      throw err
+    }
+    return Promise.resolve()
```
Comment (andrewwa-kainos) [inline src/main/app/pay/payClient.ts]: This line is unreachable.
This method should return a mutated invocation of the `patch` Promise.
```
return request.patch(...)
  .success(onFulfilment => Promise.resolve())
  .catch(onRejection => Promise.reject())
```
This also eliminates the need for the `try...catch` block.

Comment (karthikkannankainos) [reply src/main/app/pay/payClient.ts]: yeah, it works. So I have made changes as suggested above. @andrewwa-kainos

File: src/main/features/claim/routes/pay.ts (hunk 1 above)
Comment (karthikkannankainos) [reply src/main/features/claim/routes/pay.ts]: done

File: src/main/features/claim/routes/pay.ts [hunk 6]
```diff
@@ -105,6 +108,12 @@ async function successHandler (res, next) {
       )
     }
   }
+  const payClient: PayClient = await getPayClient(req)
+  const paymentReference = draft.document.claimant.payment.reference
+
+  const ccdCaseNumber = savedClaim.ccdCaseId === undefined ? 'UNKNOWN' : String(savedClaim.ccdCaseId)
+
+  await payClient.update(user, paymentReference, savedClaim.externalId, ccdCaseNumber)
```
Comment (KiranVarma02) [inline src/main/features/claim/routes/pay.ts]: Do we not have to wait for reply before redirecting? what happens if this call is not successful?

**Pull request merged**